From 1352e21d06605b0a70c0bf1ab8f1d1213dee2703 Mon Sep 17 00:00:00 2001
From: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
Date: Sat, 15 Jul 2023 20:17:04 +0530
Subject: [PATCH] wifi: ath12k: handle hw link id in Rx Error process

HW reports the hw link across the group formed in QMI. So get the radio
from the hw link id information. Otherwise all frames not handled properly
which lead to idle descriptor not in idle state.

Signed-off-by: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_rx.c | 29 +++++++++++++++----------
 1 file changed, 18 insertions(+), 11 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -4355,12 +4355,12 @@ int ath12k_dp_rx_process_err(struct ath1
 	struct hal_srng *srng;
 	struct ath12k_dp *dp;
 	void *link_desc_va;
-	int mac_id;
-	struct ath12k *ar;
+	struct ath12k *ar = NULL;
 	dma_addr_t paddr;
 	u32 *desc;
 	bool is_frag, drop = false;
 	char buf[64] = {0};
+	u8 hw_link_id;
 
 	tot_n_bufs_reaped = 0;
 	quota = budget;
@@ -4414,17 +4414,24 @@ int ath12k_dp_rx_process_err(struct ath1
 						      HAL_WBM_REL_BM_ACT_PUT_IN_IDLE);
 		}
 
-		for (i = 0; i < num_msdus; i++) {
-			mac_id = u32_get_bits(reo_desc->info0,
-					      HAL_REO_DEST_RING_INFO0_SRC_LINK_ID);
-			if (mac_id >= ab->num_radios) {
-				tot_n_bufs_reaped++;
-				ath12k_warn(ab, "invalid mac id %d on err process\n", mac_id);
-				continue;
-			}
+		hw_link_id = u32_get_bits(reo_desc->info0,
+					  HAL_REO_DEST_RING_INFO0_SRC_LINK_ID);
+
+		rcu_read_lock();
+		if (hw_link_id < ATH12K_GROUP_MAX_RADIO)
+			ar = rcu_dereference(ab->ag->hw_links[hw_link_id]);
+		rcu_read_unlock();
+
+		if (!ar || ar->ab != ab) {
+			ath12k_err(ab, "invalid src link id %d drop %d on chip id %d err process\n",
+				   hw_link_id, drop, ab->chip_id);
 
-			ar = ab->pdevs[mac_id].ar;
+			ath12k_err_dump(ab, NULL, "rx err desc: ", reo_desc,
+					sizeof(*reo_desc), srng);
+			BUG_ON(1);
+		}
 
+		for (i = 0; i < num_msdus; i++) {
 			if (drop)
 				ar->wmm_stats.total_wmm_rx_drop[ar->wmm_stats.rx_type]++;
 
